frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -498,8 +496,16 @@ func dispatchCopy(d *dispatchState, c instructions.SourcesAndDest, sourceState l
 			}
 			target := path.Join(fmt.Sprintf("/src-%d", i), f)
 			args = append(args, target)
-			mounts = append(mounts, llb.AddMount(target, llb.HTTP(src, llb.Filename(f), dfCmd(c)), llb.Readonly))
+			mounts = append(mounts, llb.AddMount(path.Dir(target), llb.HTTP(src, llb.Filename(f), dfCmd(c)), llb.Readonly))
 		} else {
+			// Resources from remote URLs are not decompressed.
+			// https://docs.docker.com/engine/reference/builder/#add
+			//
+			// Note: mixing up remote archives and local archives in a single ADD instruction
+			// would result in undefined behavior: https://github.com/moby/buildkit/pull/387#discussion_r189494717
+			if isAddCommand {
+				unpack = true
```

Comment:
wouldn't it be safer if this was so that initially `unpack = isAddCommand` and if we detect that at least one of the sources is a URL it switches unpack to `false`. I think `ADD foo.txt http://.../bar.tar.gz /` is much more legitimate than `ADD http://.../foo.tar.gz bar.tar.gz /`.

Reply:
updated